File: 704-binary-search/704-binary-search.java
class Solution {
  private int searchRec(int[] nums, int target, int lo, int hi) {
    if (lo > hi) return -1;
    int mid = (lo + hi) / 2;
    int v = nums[mid];
    if (v == target) { return mid; }
    if (v > target) { return searchRec(nums, target, lo, mid-1); }
    return searchRec(nums, target, mid+1, hi);
  }
  public int search(int[] nums, int target) {
    return searchRec(nums, target, 0, nums.length-1);
  }
}

File: 704-binary-search/704-binary-search.rs
impl Solution {
  pub fn search(nums: Vec<i32>, target: i32) -> i32 {
    let mut lo: usize = 0;
    let mut hi: usize = nums.len() - 1;
    while lo <= hi {
      let mid = (lo + hi) / 2;
      if nums[mid] == target { return mid as i32 }
      if nums[mid] > target {
        if mid == 0 { 
          return -1 
        } else {
          hi = mid - 1;
        }
      } else {
        lo = mid + 1;
      }
    }
    -1
  }
}